### ISSUE TITLE
ERL-1374: application:stop(os_mon) causes unexpected error

### ISSUE THREAD

**Issue by OTP-Maintainer** [erlang/otp#4221]
Original reporter: `paulo.ferraz.oliveira@gmail.com`
Affected version: `Not Specified`
Component: `Not Specified`
Migrated from: https://bugs.erlang.org/browse/ERL-1374

---

```
The _error_ in question seems to be coming from {{cpu_sup.c}}:
{code:java}
if(rc == 0)
  error("Erlang has closed");
{code}
Is it possible to shut {{os_mon}} down in such a way that it doesn't provoke this?
```

**Comment by rickard-green** [comment]
@paulo-ferraz-oliveira sorry for late reply. A PR is welcome!

The nicest way to shut down the ports is to send them quit commands and let them terminate without printouts when  the `os_mon` application is stopped and let them print the "Erlang has closed" error when the pipe is unexpectedly closed.

`cpu_sup` actually already has this implemented in its `stop()` function, but for some reason `os_mon` doesn't use it.

**Comment by paulo-ferraz-oliveira** [comment]
Sure, @rickard-green, no prob. I'll try to find some time for this. Thanks.

**Comment by paulo-ferraz-oliveira** [comment]
> but for some reason os_mon doesn't use it

Is it possibly it's because `application:stop(os_mon)` will already eventually call `cpu_sup:stop()` via the supervision tree?

**Comment by paulo-ferraz-oliveira** [comment]
When calling `application:stop(os_mon)` I was able to see a PING message (from `cpu_sup.c`). I didn't go down the rabbit hole, though...

**Comment by jdamanalo** [comment]
This also affects `memsup`.

Top google search for "os_mon Erlang has closed".

[[os_mon] memory supervisor port (memsup): Erlang has closed](https://elixirforum.com/t/os-mon-memory-supervisor-port-memsup-erlang-has-closed/34359)
[Suppress os_mon logging messages](https://elixirforum.com/t/suppress-os-mon-logging-messages/37212)
[[os_mon] cpu supervisor port (cpu_sup): Erlang has closed](https://github.com/vernemq/vernemq/issues/137)
[[os_mon] cpu supervisor port (cpu_sup): Erlang has closed](https://github.com/processone/tsung/issues/263)

**Comment by rickard-green** [comment]
Fixed by #6689 which has been merged into the `maint` branch which will be released in the upcoming OTP 25.3

**Comment by jonatanklosko** [comment]
Hey, I was just searching for some discussion around this error to revisit it. Whenever I stop the VM process (specifically [Livebook](https://github.com/livebook-dev/livebook), where we use os_mon to show memory usage), the "Erlang has closed" error gets printed, which I find surprising.

Is there are reason the cpu_sup/memsup OS processes don't just silently finish when the VM terminates? The error message is not really helpful, because when the whole VM terminates, it is quite self-evident. My reasoning is that, the fact that the VM outsources work to those separate OS processes doesn't matter, when the VM terminates, we just need to make sure the separate processes also terminate, it's not necessarily an error.

Thoughts?
